sdk/profiles/src/main/java/io/opentelemetry/sdk/profiles/data/ProfileContainerData.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.profiles.data;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
+import io.opentelemetry.sdk.resources.Resource;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A ProfileContainer represents a single profile. It wraps pprof profile with OpenTelemetry
+ * specific metadata.
+ *
+ * @see "profiles.proto::ProfileContainer"
+ */
+@Immutable
+public interface ProfileContainerData {
```

Comment:
I wonder if we need to differentiate between ProfileContainer and ProfileData. The proto seems to do it purely because they're extending pprof. I'm inclined to flatten this to a single class called ProfileData.

Reply:
I think perhaps this is another instance of figuring out how we want to position the SDK to users, much like the deprecated methods question.
I can appreciate the 'OpenTelemetry is the One True Way' approach of providing only functionality that does things the Otel way, but... There may also be use cases where it is desirable to e.g. generate pprof format output for feeding into pre-existing tools that expect it. For such use, being able to handle ProfileData (i.e. pprof/pprofextended) without thinking about the OTel wrapper (ProfileContainerData) makes sense. I don't really want to have to construct and pass a ProfileContainerData around when I'm only interested in the ProfileData field of it. I'm therefore slightly in favour of retaining the greater modularity and flexibility here, as I don't see it having a particularly high overhead in either runtime or maintenance and indeed the code is perhaps even simpler with it in place e.g. marshallers are more loosely coupled.

Reply:
> I don't really want to have to construct and pass a ProfileContainerData around when I'm only interested in the ProfileData field of it. 

What's the difference? There should only really be one pprof serializer ever written for this. Doesn't seem to challenging for the developers of that one component to know that we don't differentiate between the ProfileContainer and ProfileData because we don't carry the same historical baggage as the proto message definitions.

> I think perhaps this is another instance of figuring out how we want to position the SDK to users, much like the deprecated methods question.

Agreed. I like this a heuristic (or part of one anyway) could be that as long as its possible to do a pprof encoding of the data, we can be free to have some minor divergences from the proto representation.

> A 1:1 between data classes and marshallers helps modularity and testing.

That could be true, but not obvious to me right now whether it is or isnt. 

Overall, I don't feel strongly one way or another about this. I doubt this is the last word we'll have on the matter, so no need to split hairs. 🙂

Reply:
> There should only really be one pprof serializer ever written for this.

Hmm. There is an existing ecosystem of tools written to handle pprof files and retaining compatibility in the proto format is aimed at allowing us to take advantage of those. Thus I think there may be use cases where you want to export pprof to file, which means writing a ProfileData without the enclosing ProfileContainer.

So yes, technically still only one Marshaler for a ProfileData, but it has got two use cases - to be called as an intermediate node in a tree that's rooted at the ExportProfilesServiceRequest and works down via the ProfileContainer, which is how OTLP wire messages will get written, or to be called as a root node directly, which is how pprof files would get written.

I'd like to retain pprof file export capability at least through the experimental phase, as it's useful to be able to e.g. send into a pprof flame graph rendering tool to visually sanity check the encoding. If it's something we want to offer in the eventual stable API is a different question and even then the pipeline design may be such that it winds up being cleaner to create the full tree in all cases but ignore the container bits in the pprof file exporter backend.

Reply:
I meant there should only be one pprof serializer - of course there is also the need for an OTLP serializer. My point was that removing the additional abstraction layer should only marginally impact the pprof serializer. And since it only needs to be implemented once, its not the end of the world if the developer who writes has to understand the difference between the opentelemetry-java in-memory representation and pprof.

> I'd like to retain pprof file export capability at least through the experimental phase

Sure, I don't feel strongly one way or the other.